crates/arroyo-connectors/src/nats/mod.rs

```diff
@@ -0,0 +1,316 @@
+        });
+    }
+
+    fn from_config(
+        &self,
+        id: Option<i64>,
+        name: &str,
+        config: NatsConfig,
+        table: NatsTable,
+        schema: Option<&ConnectionSchema>,
+    ) -> anyhow::Result<Connection> {
+        let stream_or_subject = match &table.connector_type {
+            ConnectorType::Source { stream, .. } => stream.clone(),
+            ConnectorType::Sink { subject, .. } => subject.clone(),
+        };
+
+        let (connection_type, desc) = match table.connector_type {
+            ConnectorType::Source { .. } => (
+                ConnectionType::Source,
+                format!("NatsSource<{:?}>", stream_or_subject.clone()),
+            ),
+            ConnectorType::Sink { .. } => (
+                ConnectionType::Sink,
+                format!("NatsSink<{:?}>", stream_or_subject.clone()),
+            ),
+        };
+
+        let schema = schema
+            .map(|s| s.to_owned())
+            .ok_or_else(|| anyhow!("No schema defined for NATS connection"))?;
+
+        let format = schema
+            .format
+            .as_ref()
+            .map(|t| t.to_owned())
+            .ok_or_else(|| anyhow!("'format' must be set for NATS connection"))?;
+
+        let config = OperatorConfig {
+            connection: serde_json::to_value(config).unwrap(),
+            table: serde_json::to_value(table).unwrap(),
+            rate_limit: None,
+            format: Some(format),
+            bad_data: schema.bad_data.clone(),
+            framing: schema.framing.clone(),
+        };
+
+        Ok(Connection {
+            id,
+            connector: self.name(),
+            name: name.to_string(),
+            connection_type: connection_type,
+            schema,
+            config: serde_json::to_string(&config).unwrap(),
+            description: desc,
+        })
+    }
+
+    fn from_options(
+        &self,
+        name: &str,
+        options: &mut HashMap<String, String>,
+        schema: Option<&ConnectionSchema>,
+        profile: Option<&ConnectionProfile>,
+    ) -> anyhow::Result<Connection> {
+        let connection = profile
+            .map(|p| {
+                serde_json::from_value(p.config.clone()).map_err(|e| {
+                    anyhow!("Invalid config for profile '{}' in database: {}", p.id, e)
+                })
+            })
+            .unwrap_or_else(|| Self::connection_from_options(options))?;
+
+        let table = Self::table_from_options(options)?;
+
+        Self::from_config(&self, None, name, connection, table, schema)
+    }
+
+    fn make_operator(
+        &self,
+        profile: Self::ProfileT,
+        table: Self::TableT,
+        config: OperatorConfig,
+    ) -> anyhow::Result<OperatorNode> {
+        Ok(match table.connector_type {
+            ConnectorType::Source { .. } => OperatorNode::from_source(Box::new(NatsSourceFunc {
+                stream: table.client_configs.get("nats.stream").cloned().unwrap(),
```

Comment:
This is expecting that client_configs will contain a `"nats.stream"` key, but when creating through the web ui that won't necessary be there. Instead, you can use the stream argument for the source like `ConnectorType::Source { stream }`.

We also want to avoid unwraps (and other panicking constructs) in this code, and instead return an error when the data isn't what we expect.

Reply:
Done 👍